improve support for parallel file format changes

wsu-cb opened the pull request:
Note:  This might be a wild idea, so I have set it as a draft, and I have set the file version number to 105.0.

Using a file version number makes it difficult for developers to make file format changes in parallel.  For example, in previous work, it was common for test files to have the same version number, but different formats.  This makes it difficult to keep a stock of test files.

Therefore, modify the file format to allow files to more clearly describe their format.

DLLarson commented:
Hi Bill,

I've done cursory scan of the code changes.

Could you do me a favor and add a summary of the theory of operation of this change to `versions.h` file? I want make sure I grok what's going on here and how the newer version handling is evolved from the former. I'm pretty sure I could figure it out by careful study of the code changes but this mod is too important to count on my interpreting the changes correctly. A couple of paragraphs would help a lot.

-Dale

wsu-cb commented:
The goal of this PR is to make one final file version NUMBER change, and then never change the file version NUMBER again.  The problem with version numbers is there isn't a good way to make file format changes in parallel, and, at least as we were using them, no good way to distinguish the evolution of the file format between releases.  This made it difficult to read files written at intermediate points in time between releases because the version number was the same, but the actual format was different.

I propose replacing the storage-efficient but not-very-informative version NUMBER with a much-larger storage but more informative version FEATURE LIST that a particular file requires its reader to support.  There is also a requirement that readers that know about features NOT specified by a file will correctly interpret that file in a backward-compatible pre-feature mode.

A "feature" has value semantics (https://en.wikipedia.org/wiki/Value_semantics), and is represented by a name of string type.  It is assumed that two developers are unlikely to pick the same string name for two features at near enough the same time to represent a new conflict in file representation, but if that possibility is a concern, we could have a policy that all feature names be prefixed with the GitHub user ID of the developer.  It is of course each developer's responsibility not to use the same feature name for different purposes.  Because I think it is more convenient (see below), I have decided that the string names are case-insensitive.  These names are represented (in `Version.h`) by class `Feature`.  `Feature` is implemented as a thin wrapper on `std::string` that provides the very small interface needed to represent a feature (identity comparison, and save/load).

A collection of `Feature`s is represented by class `Features`.  Again, this has value semantics and is just a thin wrapper on `std::vector<Feature>` providing the very small interface needed (add/remove feature, test whether it contains a feature, and save/load).

The baseline, no features, file is almost the same as file version 3.10.  The only difference is the change in version number, and the addition of the empty feature list.

As a demonstration of the concept, I have re-implemented file version 4.0 using this design.  File version 4.0 added several features to the system, which are contained in the `Features` returned by `GetCBFile4Features()`.  So a file with the feature list in `GetCBFile4Features()` is almost the same as file version 4.0.  The only difference is the change in version number, and the addition of the `GetCBFile4Features()` feature list.

Random notes:

File format 4.0 included a change in the file representation of count information.  Both feature name length and feature list length are counts, but the reader can't know the count representation until after reading the feature list.  So, the file format for `Feature` and `Features` must have a fixed format independent of the file version.

I added an RAII helper `SetFileFeaturesGuard` to parallel the pre-existing version RAII helper `SetLoadingVersionGuard`. 

Another advantage of this design is that it's easier to automatically maintain compatibility with previous versions of CB.  For instance, the code in this PR only writes files using 32bit ID values when there is at least one ID that can't be represented in 16bits.  So if we had already been using this design when we added 32bit ID support, we would have automatically been saving the file as compatible with previous CB versions unless we actually needed the expanded file storage.

On the other hand, for testing purposes, I have added command line switch support to force using a file feature even if it isn't actually required, e.g., `/filever:force-id-32bit`.  Also, I have added support to forbid saving a file with a particular feature, e.g., `/filever:no-id-32bit`.  This is only checked at save time, not continuously during editing, but at least the user can verify that he hasn't unintentionally written a file that is no longer compatible with another player's CB.  For convenience of typing these command line switches, feature names are case-insensitive.  (The pre-existing `/filever:3.10` and `/filever:4.0` still exist.)

I fixed some missing special case handling for 64bit processes writing `size_t` in `/filever:3.10`.

DLLarson commented:
>Don't read yet; more to come...

Ok... 

If this ultimately comes to pass this info should be placed directly in the source code.

-Dale

wsu-cb commented:
Unfortunately, I can't find a way to create links to specific lines of code in the PR.  If I could, I would put some links to the corresponding code in https://github.com/CyberBoardPBEM/cbwindows/pull/115#issuecomment-1457521424.

DLLarson commented:
> Unfortunately, I can't find a way to create links to specific lines of code in the PR. If I could, I would put some links to the corresponding code in [#115 (comment)](https://github.com/CyberBoardPBEM/cbwindows/pull/115#issuecomment-1457521424).

I did find this:

https://stackoverflow.com/questions/23821235/how-to-link-to-a-specific-line-number-on-github

It describes several ways to create an URL to line(s) in a source code file or specifically [to some lines of code in the PR like this](https://github.com/CyberBoardPBEM/cbwindows/pull/115/files#diff-7cd01086ad22b03aa6a0a614d1b5414d441218a3d4456eab1954e17623f2865fR567).

A link to GitHub's docs for the same: https://help.github.com/en/articles/getting-permanent-links-to-files

-Dale

DLLarson commented:
I've read through your description and it's interesting. I want to spend some time studying it to feel it out. 

One question that popped into my head was: Can the file load/save code load features it doesn't recognize and reconstitute them when saving? The idea being they wouldn't be lost even if they weren't meaningful to an earlier shipped or parallel development version of CB?


-Dale

wsu-cb commented:
> A link to GitHub's docs for the same: https://help.github.com/en/articles/getting-permanent-links-to-files
> 
> -Dale

Good to know in the future.  Thank you.

wsu-cb commented:
> I've read through your description and it's interesting. I want to spend some time studying it to feel it out.
> 

As a "wild idea," I would expect it to require some time to digest.

> One question that popped into my head was: Can the file load/save code load features it doesn't recognize and reconstitute them when saving? The idea being they wouldn't be lost even if they weren't meaningful to an earlier shipped or parallel development version of CB?
> 
> -Dale

As currently implemented, no, because there's not enough information in the file to enable determining where features begin and end, much less which feature is about to begin.  I haven't given it serious thought, but my initial reaction is that, with a substantially greater change to the file format, we could construct something roughly analogous to XML, which has explicit feature identifiers and begin and end markers, and that would allow what I think you're describing.  (Or even convert the file format to XML.  Personally, I dislike text file formats for largely binary data, but it can be done.)

However, even if a reader could find the parts of the file it understood, the end result would not be usable in general.  Consider the example of the change to 32-bit IDs:  If a reader doesn't understand that feature, it also wouldn't have the updated storage tables to handle 32-bit IDs, and there would be too many IDs in the file to be handled by the 16-bit storage tables.  So the reader wouldn't be able to give objects their IDs, and that just doesn't seem like a viable state to support.  Likewise, 6-sided pieces would be unusable to a 2-sided piece CB.  There are some examples, such as saving `CRollState`, of truly optional data, but I suspect that is more the exception than the typical case.  I'm afraid I think trying to support partial file reads would require more effort than the benefit warrants.

DLLarson commented:
Just an update...

It's been slow going on digging into this the past week. I hope to devote more time this coming week.

-Dale

wsu-cb commented:
> It's been slow going on digging into this the past week. I hope to devote more time this coming week.
> 
> -Dale

Thank you for the update.

DLLarson commented:
>Thank you for the update.

FWIW...I'm getting closer finishing this review. I expect to be done before the weekend is out.

-Dale

DLLarson reviewed (commented):
Hi Bill,

I'm done looking and playing around with this new code.

It looks like a really good idea.

In a nutshell, basically, the file versioning scheme boils down to replacing version numbers with feature names while keeping the existing code techniques we used to archive the actual feature's data.

Some questions and observations...

1. For an existing feature in the new scheme, how would a feature revision look? (Like perhaps some flags needed to be added to it.)

2. When a version of the *program* (not file structure) is released will the release notes associate newly defined features with a particular program version?

3. It looks like a table of features will be defined in the program for the current supported features. Considering item (2) above, it would be good if the table is filled oldest feature to newest added feature order and divided into sections identified by comments stating which version of the *PROGRAM* added the features. This is kinda nice since a developer can get a quick look at "when" a feature was added.

4. The longer I explored the implementation I found that it looks much like a command line option processing facility except Features only have a value of existence or non-existence (not assigning booleans, numbers, strings, etc...) because you provide a way to inject/enable/disable them. It doesn't take much imagination to see this feature can be used to activate other things such as diagnostic code (for instance) that has nothing to do file formats.

A side note.... I think the copyright end years and attributions should be updated in any file headers that have new code added to them.


-Dale

wsu-cb commented:
> Hi Bill,
> 
> I'm done looking and playing around with this new code.
> 
> It looks like a really good idea.
> 

I'm glad you approve of the concept.



> In a nutshell, basically, the file versioning scheme boils down to replacing version numbers with feature names while keeping the existing code techniques we used to archive the actual feature's data.
> 

Exactly.



> Some questions and observations...
> 
> 1. For an existing feature in the new scheme, how would a feature revision look? (Like perhaps some flags needed to be added to it.)

I think it would be on a case-by-case basis.  Basically, whatever makes it reasonably clear.  I could imagine some cases being something like `private-player-board-v2` and other cases being something like `size_t-64bit-with-compression`.



> 2. When a version of the _program_ (not file structure) is released will the release notes associate newly defined features with a particular program version?

I think that's an orthogonal question to what this PR is addressing, isn't it?  In other words, wouldn't the same question be equally valid if we continued to use file version numbers instead of feature lists to describe the file format?  It seems reasonable for the release notes to say that CB4.0 added a new private player boards feature that provides complete hidden movement.  My own feeling is that listing that this new feature is serialized in file format version `4.2` or `class Feature("private-player-boards")` is an implementation detail that's too detailed for release notes.  On the other hand, if we want the release notes to be that detailed, they can be, no matter how we internally represent the file version info.



> 3. It looks like a table of features will be defined in the program for the current supported features. Considering item (2) above, it would be good if the table is filled oldest feature to newest added feature order and divided into sections identified by comments stating which version of the _PROGRAM_ added the features. This is kinda nice since a developer can get a quick look at "when" a feature was added.

I usually think of feature lists running newest to oldest, rather than oldest to newest, but I don't have strong feelings on the subject.



> 4. The longer I explored the implementation I found that it looks much like a command line option processing facility except Features only have a value of existence or non-existence (not assigning booleans, numbers, strings, etc...) because you provide a way to inject/enable/disable them. It doesn't take much imagination to see this feature can be used to activate other things such as diagnostic code (for instance) that has nothing to do file formats.
> 

I'm having trouble working out what you're saying here.  I certainly agree that command line options can specify any type of data that is both conveniently typed by the user on a keyboard and conveniently parsed by the program at startup.  Many programs have command line features to control aspects of the program unrelated to file format.  GCC has a list of command line options that goes on for pages.  I certainly don't object to adding more command line options to CB if they seem useful.



> A side note.... I think the copyright end years and attributions should be updated in any file headers that have new code added to them.
> 

I Am Not A Lawyer.  What exactly do you want the copyright line(s?) to say?  And what exactly is the difference between "new code" and "changed code"?


> -Dale

DLLarson commented:
>I think that's an orthogonal question to what this PR is addressing, isn't it? 

Yes. Ignore it.

>I'm having trouble working out what you're saying here.

All I'm saying is that your *implementation* would be useful for controlling simple diagnostic "features" that aren't file format related. Nothing deeper than that. I also learned more about STL regex support. Pretty dang powerful.

Anyway...

I'm good with merging this PR into the mainline if you're comfortable with it.

-Dale

DLLarson commented:
>I Am Not A Lawyer. 

All I'm saying here is change the end date of the file's copyright notice to match our current year if we made non-trivial changes to them. To wit...
```
// Copyright (c) 1994-2020 By Dale L. Larson, All Rights Reserved.
                      vvvv                   vvvvvvvvvvvvv       
// Copyright (c) 1994-2023 By Dale L. Larson & William Su, All Rights Reserved.
```
or however you'd like to declare your contributions--which have been substantial.

-Dale

wsu-cb commented on GP/GamDoc3.cpp:
```diff
@@ -453,12 +586,63 @@ void CGamDoc::SerializeScenario(CArchive& ar)
+            }
+        }
+        else if (NumVersion(fileGsnVerMajor, fileGsnVerMinor) == NumVersion(4, 0) ||
+            NumVersion(fileGsnVerMajor, fileGsnVerMinor) == NumVersion(104, 5)) {
```
I discovered this problem and fixed it in commit fafae6.  (I had failed to clean out all of my experimental code.)  But as for version 105, see below.

wsu-cb commented on GP/GamDoc3.cpp:
```diff
@@ -113,11 +114,73 @@ void CGamDoc::SerializeMoveSet(CArchive& ar, CHistRecord*& pHist)
+            /* these features are global in the sense that, if
+                any ID needs 32 bits, then we store all IDs in
+                32 bit format */
+            if (GetBoardManager()->Needs32BitIDs() ||
```
I THINK the comment is true, but it wouldn't be the first time I made a mistake.  Can you point me to the specific area(s) where you think the ID sizes vary?

DLLarson replied:
I cant seem to find the code that triggered my above comment. Next time I run into something like that I'll add a link to the code that the caused the confusion.

wsu-cb commented:
> I'm good with merging this PR into the mainline if you're comfortable with it.
> 
> -Dale

Actually, I don't want to merge this just yet.

1) I am inclined to switch the version number to 5.0.  I was using version 105.0 in case you didn't like this concept, and we someday used version 5.0 for something else.

2)  You have made a couple of comments above that we should resolve.

3) You previously said you would like some additional explanation added to the code.

wsu-cb commented:
I assume this would be a separate PR.

DLLarson commented:
> > All I'm saying here is change the end date of the file's copyright notice to match our current year if we made non-trivial changes to them.
> 
> I assume this would be a separate PR.

Actually no. Changes to copyright dates go hand-in-hand with change of content. They aren't separate things.

-Dale

DLLarson commented:
>I don't want to merge this just yet.

OK... I got fatigued and forgot about those.

DLLarson commented:
Just a warning. My access using SSH to get your current changes is busted on GitHub's side. After digging around I found this like explaining what is going on:

https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/

-Dale

DLLarson reviewed (changes requested):
Missed a copyright.

Question. I ran and upgraded old gamebox and game and the only features is see in the files with a hex dumper is `size_t-64bit` for the GBX. For the GAM I only see `size_t-64bit` and `CRollState`. What will activate using the full feature set you've added?

-Dale

wsu-cb commented on GShr/Versions.h:
```diff
@@ -1,14 +1,140 @@
 //
 // wsu20210731
 //      4.00 - 32-bit TileID/MarkID/PieceID/BoardID
+inline const Feature ftrId32Bit("id-32bit");
```
Thank you.

wsu-cb commented on GShr/WinState.cpp:
As stated in https://github.com/CyberBoardPBEM/cbwindows/pull/115/commits/874bcd1b8271e016d6b4617771ed25af4e3da7d2, I only changed copyright of files with >=10 changes this year as reported by `git diff --stat`, which reports 8 changes for `GShr/WinState.cpp`.  Are you suggesting a different policy?

DLLarson replied:
Raw quantity of changes is not a good measure. If the code has changed then update the copyright. I would say a fix of some misspelled comments would be a reason not to do it. When in doubt, just change the copyright. Note that updating the copyright is a benefit to you to recognize your contributions. It doesn't make sense to avoid doing it.

wsu-cb commented:
> Question. I ran and upgraded old gamebox and game and the only features is see in the files with a hex dumper is `size_t-64bit` for the GBX. For the GAM I only see `size_t-64bit` and `CRollState`. What will activate using the full feature set you've added?
> 

For `"id-32bit"`, you can use `/filever:force-id-32-bit` to force the system to use it, even with a small set of objects, or you can create a `.gbx` with 66,000 pieces (probably via temporary code using a loop to repeated create many instances of a piece).

For `"geo-square-cell"`, you would have to create a `.gsn` with an actual geomorphic board with square cells.   (You could use `/filever:force-square-cell` to cause the `Feature` to appear in the feature list at the end of the file, but without a geoboard actually using square cells, you would not find anything interesting in the data.)

For `"geo-rotate-unit"`, if you use `/filever:force-geo-rotate-unit`, all geoboards will be stored with a `Rotation90` value (which might be `Rotation90::r0`), but you would still at least need a geoboard, not just standard boards, to see anything interesting.  Or you could create a geoboard with a rotated unit board.

"piece-100-sides" and "private-player-board" are similar to `"id-32bit"`; they have storage implications even for pieces/boards not actually requiring the capability.

DLLarson commented:
> For "id-32bit", you can use ... for pieces/boards not actually requiring the capability.

Wouldn't we want them all to be active since they are new features we want to use by default when we ship product? I can see the need for manual activation in initial testing but past that it doesn't make sense.

I hope I'm not missing something here.

-Dale

wsu-cb commented:
> > For "id-32bit", you can use ... for pieces/boards not actually requiring the capability.
> 
> Wouldn't we want them all to be active since they are new features we want to use by default when we ship product? I can see the need for manual activation in initial testing but past that it doesn't make sense.

Unless `"/filever:no-id-32bit"` has been set, the 32bit-id capability IS in effect, so we are using it by default.  The distinction is how the file format represents the data, not whether it's in effect.  The `"id-32bit"` file format `Feature` is only NECESSARY when there are more ids (in a single table) than 16bits can represent.  Therefore, if only 16bits are NECESSARY when writing the file, I use the older file format so that older CB users don't need to upgrade.  (It also provides a slight file compression, but that's just a nice bonus, not an actual design goal, and might not be true for other `Feature`s.)  Or, to be entirely precise, that's what I would have done if I had thought of this design before making the changes that created file format 4.0.

This is the distinction between the `GetCBFeatures()` list (which are the `Features` available to put in a `CArchive`) and `GetFeatures(const CArchive& ar)` (which is the subset of `GetCBFeatures()` that the `Serialize` code has actually put in the `CArchive` (plus whatever /filever:force-* is requiring to be the `CArchive`, even if not actually necessary to fully represent the data)).

DLLarson commented:
Thanks for the clarification. I figured I was missing something. Great stuff!

When you're ready give me your thumb's up we'll merge. Right now it's still tagged as a work in progress.

-Dale

DLLarson reviewed (approved):
It looks good to go!

-Dale

DLLarson commented:
I would like to do a squash commit with this PR if you don't object.

My reason is that the composite of commits on this particular PR really add up to a single commit idea. Squashing should provide a more coherent commit operation rather than pieces spread about in several commits.

-Dale

wsu-cb commented:
> I would like to do a squash commit with this PR if you don't object.
> 
> My reason is that the composite of commits on this particular PR really add up to a single commit idea. Squashing should provide a more coherent commit operation rather than pieces spread about in several commits.

I agree.
